### PR TITLE
[FEAT] precompile-sha256

### PR DESCRIPTION
**Pull request by noel2004**
This PR complete the precompile SHA256

The SHA256 circuit made used of `table16` code in `halo2_gadget`, by a forking in current repo

### known issue:

- [ ] Initialize state is not constrainted in `table16`
- [ ] The `SHA256Instructions` trait is not suitable for the SHA-2 protocol since it is not allowed chips recognizing the final block (some constraints for padding is required)
- [ ] The circuit cost too much rows for a 512 bit block (>2000)

## Breaking change

The dependency for L2 geth in `geth-util` has been updated to enable precompile SHA256

### CODE REVIEW

**Comment by noel2004** [comment]
\b run testool 1

File: bus-mapping/src/evm/opcodes/precompiles/mod.rs [hunk 1]
```diff
@@ -50,6 +52,23 @@ pub fn gen_associated_ops(
...
+                    digest: output_bytes
+                        .try_into()
+                        .expect("output bytes must be 32 bytes"),
+                }))
+            },
+            Some(PrecompileAuxData::Base {
+                input_bytes: input_bytes.to_vec(),
+                output_bytes: output_bytes.to_vec(),
+                return_bytes: return_bytes.to_vec(),
+            }),
```
Comment (roynalnaruto) [inline bus-mapping/src/evm/opcodes/precompiles/mod.rs]: I will suggest we refactor this into `PrecompileAuxData::Sha256` just to avoid using any component related to the `Base` case (precompiles that are not supported), similar to what we do for the `Identity` case.

File: zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+        precompile::PrecompileCalls,
+    };
+    use eth_types::{bytecode, ToWord};
+    use itertools::Itertools;
+    use mock::TestContext;
+
+    use crate::test_util::CircuitTestBuilder;
+
+    lazy_static::lazy_static! {
+        static ref TEST_VECTOR: Vec<PrecompileCallArgs> = {
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs:224]: We should have more test cases:
1. calldata_length > 32
2. calldata_length == 0
3. returndata_size < 32
4. returndata_offset < calldata_offset + calldata_length
5. calldata_length is a multiple of 32
6. out of gas case (to see that it is appropriately handled by the OOG gadget)

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs [hunk 3]
```diff
@@ -1476,6 +1476,22 @@ impl<'a, F: Field> EVMConstraintBuilder<'a, F> {
...
+        &mut self,
+        input_rlc: Expression<F>,
+        output_rlc: Expression<F>,
+    ) {
+        self.add_lookup(
+            "sha256 lookup",
+            Lookup::Sha256Table {
+                input_rlc,
+                output_rlc,
+            },
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/util/constraint_builder.rs]: How does the SHA256 lookup differentiate between:
1. SHA256([0, 0])
2. SHA256([0])

The `input_rlc` for the above is the same, however the output should be different. So we also need to take into account the `input_len` (length of the input bytes) if I am not wrong?

**Review by roynalnaruto — CHANGES_REQUESTED** [review]
Reviewed the EVM circuit related parts.

File: zkevm-circuits/src/evm_circuit/util/constraint_builder.rs (hunk 3 above)
Comment (noel2004) [reply zkevm-circuits/src/evm_circuit/util/constraint_builder.rs]: now resolved

File: zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs [hunk 4]
```diff
@@ -78,7 +78,11 @@ impl<F: Field> ExecutionGadget<F> for SHA256Gadget<F> {
 
         // sha256 verify lookup
         cb.condition(is_success.expr(), |cb| {
-            cb.sha256_table_lookup(input_bytes_rlc.expr(), output_bytes_rlc.expr());
+            cb.sha256_table_lookup(
+                input_bytes_rlc.expr(),
+                call_data_length.expr(),
+                output_bytes_rlc.expr(),
+            );
```
Comment (roynalnaruto) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs:85]: We should add a test case with `call_data_length == 0`, i.e. `input_len == 0`. In this case, the precompiled contract call will be successful, so a lookup to the keccak table will be done. However, we must ensure this "empty" input case is handled from the SHA256 circuit (that it was populated as witness, and not ignored)

File: bus-mapping/src/evm/opcodes/precompiles/mod.rs (hunk 1 above)
Comment (noel2004) [reply bus-mapping/src/evm/opcodes/precompiles/mod.rs]: done

File: zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs (hunk 2 above)
Comment (noel2004) [reply zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs:224]: more case is added. However, calldata length for some edge cases (like multiple of 32) may not work as it expected since the SHA256 circuit is not involved in precompile's unittest (sha256 is filled by `dev_load`).

There are tests in repo of SHA256 circuit for hashed length with edge cases

File: zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs (hunk 4 above)
Comment (noel2004) [reply zkevm-circuits/src/evm_circuit/execution/precompiles/sha256.rs:85]: done with a dedicated unittest

File: bus-mapping/src/precompile.rs [hunk 5]
```diff
@@ -34,7 +34,7 @@ pub(crate) fn execute_precompiled(
                 // so we need to override the behavior of invalid input
                 match PrecompileCalls::from(address.0[19]) {
                     PrecompileCalls::Blake2F
-                    | PrecompileCalls::Sha256
+                    // | PrecompileCalls::Sha256
```
Comment (roynalnaruto) [inline bus-mapping/src/precompile.rs]: we can remove this comment, since post this PR SHA256 will be enabled.

**Review by roynalnaruto — COMMENTED** [review]
Looks good. Although I have not reviewed the `sha256` crate itself. Only reviewed the integration into EVM circuit precompiles.

File: bus-mapping/src/precompile.rs (hunk 5 above)
Comment (noel2004) [reply bus-mapping/src/precompile.rs]: done

**Comment by lispc** [comment]
i think you need to update the min_num_rows function of SuperCircuit?

File: zkevm-circuits/src/sha256_circuit.rs [hunk 6]
```diff
@@ -123,8 +126,27 @@ impl SubCircuit<Fr> for SHA256Circuit<Fr> {
...
+            if hasher.blocks() > self.1 {
+                log::error!("handled 512-bit block exceed limit ({})", self.1);
+                return Err(Error::Synthesis);
+            }
+        }
+
+        // paddings
+        for _i in hasher.blocks()..self.1 {
+            hasher.update(layouter, chng, &[])?;
+            hasher.finalize(layouter, chng)?;
```
Comment (lispc) [inline zkevm-circuits/src/sha256_circuit.rs:150]: will the padding be slow?  I remembered previously, we re-calculate the witness for keccak("") repeatedly.  Quite slow.  Later we first calculate the witness values only once, and then just write them into cells.  Much faster.

**Comment by noel2004** [comment]
> i think you need to update the min_num_rows function of SuperCircuit?

There is updating for [`min_num_rows_block_subcircuits` in SuperCircuit](https://github.com/scroll-tech/zkevm-circuits/blob/e3d13338cd12e103cccbe7f72c8e0629d9db79ff/zkevm-circuits/src/super_circuit.rs#L509), does it the required updates?

File: zkevm-circuits/src/sha256_circuit.rs (hunk 6 above)
Comment (noel2004) [reply zkevm-circuits/src/sha256_circuit.rs:150]: Have made a quick bench with `k=19`, by calculating only 1x 512-bit block and 128x blocks, the total proving time is:

+ 128x blocks: 17.355s
+ 1x block: 16.793s

So the overhead for synthising is ~3%, i think it is trivial so such performance enhancement is not critical.

I guess the reasons may be:

1. The number of sha256 is small and "data density" is sparse (it need 2k rows for a 512-bit block)
2. Our circuit do not need extra calculation for the input bytes.

**Comment by lispc** [comment]
\b run testool 1

**Comment by noel2004** [comment]
bump halo2 version. Since there is no too much things left for the `sha256` dir, we have merge it into zkevm-circuits

**Review by lispc — APPROVED** [review]
Can be merged After ci passes

File: zkevm-circuits/src/sha256_circuit/circuit.rs [hunk 7]
```diff
@@ -0,0 +1,1249 @@
...
+
+/// u32 size for SHA256 digit
+pub const DIGEST_SIZE: usize = 8;
+
+/// the defination for a sha256 table
+pub trait SHA256Table {
+    /// the cols has layout [s_enable, input_bytes, input_len, hashes, effect]
+    fn cols(&self) -> [Column<Any>; 5];
+
+    /// ...
```
Comment (roynalnaruto) [inline zkevm-circuits/src/sha256_circuit/circuit.rs]: Generally we should avoid adding `/// ...` simply to fix `cargo doc` warnings/failures. It should take not much time to add a small doc regarding the method. Its a nitpick but it helps a lot our code quality and readability.

**Review by roynalnaruto — APPROVED** [review]
Minor point, but otherwise looks good.

File: zkevm-circuits/src/sha256_circuit/circuit.rs (hunk 7 above)
Comment (noel2004) [reply zkevm-circuits/src/sha256_circuit/circuit.rs]: done
